### DISCUSSION TITLE
Error: P1017: Server has closed the connection. when a group of commands

### DISCUSSION THREAD

**Discussion by ElBaraaSamir** [prisma/prisma#25352] [Q&A]
### Question

I'm following a course, [here is the link](https://www.youtube.com/watch?v=GHTA143_b-s&t=9805s&ab_channel=freeCodeCamp.org)

Halfway through the course he started testing the code but made another database connection and created a new .env name it .env.test.

in .env

```
PORT=3002
JWT_SECRET=secret
DB_USERNAME=root
DB_PASSWORD=MySQL.Bebon3232
DB_HOST=db_test
DB_NAME=learning_nestjs
DATABASE_URL="mysql://root:MySQL.Bebon3232@localhost:3307/learning_nestjs"
```

in .env.test

```
PORT=3002
JWT_SECRET=secret
DB_USERNAME=root
DB_PASSWORD=MySQL.Bebon3232
DB_HOST=db_test
DB_NAME=learning_nestjs
DATABASE_URL="mysql://root:MySQL.Bebon3232@localhost:3308/learning_nestjs"
```

The only difference is the port number.

while in docker-compose 

```
version: '3.8'

services:
  dev_db:
    image: mysql:latest
    restart: always
    ports:
      - 3307:3306
    environment:
      - MYSQL_ROOT_USER=${DB_USERNAME}
      - MYSQL_ROOT_PASSWORD=${DB_PASSWORD}
      - MYSQL_DATABASE=${DB_NAME}
    networks:
      - bebon

  test_db:
    image: mysql:latest
    restart: always
    ports:
      - 3308:3306
    environment:
      - MYSQL_ROOT_USER=${DB_USERNAME}
      - MYSQL_ROOT_PASSWORD=${DB_PASSWORD}
      - MYSQL_DATABASE=${DB_NAME}
    networks:
      - bebon

networks:
  bebon:
```

When running the code everything works just fine, untell the testing part.

This is the scripts for running the code.

```
"prisma:dev:deploy": "prisma migrate deploy",
"db:dev:up": "docker compose up dev_db -d",
"db:dev:rm": "docker compose rm dev_db -s -f -v",
"db:dev:restart": "yarn db:dev:rm && yarn db:dev:up && yarn prisma:dev:deploy",

"prisma:test:deploy": "dotenv -e .env.test -- prisma migrate deploy",
"db:test:up": "docker compose up test_db -d",
"db:test:rm": "docker compose rm test_db -s -f -v",
"db:test:restart": "yarn db:test:rm && yarn db:test:up && yarn prisma:test:deploy",

"build": "nest build",
"start": "nest start",
"start:dev": "nest start --watch",

"pretest:e2e": "yarn db:test:restart",
"test:e2e": "dotenv -e .env.test -- jest --watchAll --no-cache --config ./test/jest-e2e.json"
```

I run test:e2e which runs the following commands by order before it runs itself
1. yarn compose rm test_db -s -f -v
2. yarn compose up test_db -d
3. dotenv -e .env.test -- prisma migrate deploy

And it stops when Prisma deploys and shows this error.

```
yarn test:e2e
yarn run v1.22.22
$ yarn db:test:restart
$ yarn db:test:rm && yarn db:test:up && timeout 1 && yarn prisma:test:deploy
$ docker compose rm test_db -s -f -v
[+] Stopping 1/1
 ✔ Container learning-nest-js-test_db-1  Stopped                                                                                                                                           2.0s 
Going to remove learning-nest-js-test_db-1
[+] Removing 1/0
 ✔ Container learning-nest-js-test_db-1  Removed                                                                                                                                           0.1s 
$ docker compose up test_db -d
[+] Running 1/1
 ✔ Container learning-nest-js-test_db-1  Started                                                                                                                                           0.6s 

Waiting for 0 seconds, press a key to continue ...
$ dotenv -e .env.test -- prisma migrate deploy
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": MySQL database "learning_nestjs" at "localhost:3308"

Error: P1017: Server has closed the connection.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Even though running the commands separately

    npx dotenv -e .env.test -- prisma migrate deploy

    npx dotenv -e .env.test -- jest --watchAll --no-cache --config ./test/jest-e2e.json

work just fine alone but not as a group of commands
the same goes for both of the commands 

    "db:test:restart": "yarn db:test:rm && yarn db:test:up && yarn prisma:test:deploy"

    "db:dev:restart": "yarn db:dev:rm && yarn db:dev:up && yarn prisma:dev:deploy"

It stops the containers, deletes them, and runs new ones and when it should deploy, it shows the error P1017.

It should pass the test that has been made, which is 

```
import { INestApplication, ValidationPipe } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { AppModule } from '../src/app.module';

describe('app e2e', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleRef.createNestApplication();
    app.useGlobalPipes(
      new ValidationPipe({
        whitelist: true,
      }),
    );
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  it.todo('should pass');
});

```

Note that the database containers (dev_db & test_db) work completely.

**Comment by ludralph** [comment]
Hi @ElBaraaRadwan 👋 

Have you tried running each command in isolation to make sure the database is fully initialized also how long does the MySQL container typically take to start up when using docker-compose? Does the error occur every time you run the tests, or is it intermittent?

**Reply by ElBaraaSamir** [reply]
thanks for the reply @ludralph 

- yes, every command works perfectly fine on its own, but not as a group of commands that contain prisma deploy like:-
`"db:test:restart": "yarn db:test:rm && yarn db:test:up && yarn prisma:test:deploy"`
`"db:dev:restart": "yarn db:dev:rm && yarn db:dev:up && yarn prisma:dev:deploy"`
`"pretest:e2e": "yarn db:test:restart"`

all of them result to an error. but if I run any one of them alone it will run without a problem, even prisma deploy.
-  it takes a second for the container to start up again
- no, the test command runs just fine but there is a pre-command "pretest:e2e"
`"db:test:restart": "yarn db:test:rm && yarn db:test:up && yarn prisma:test:deploy"`
which runs ok until deploying prisma it gives an error.

**Reply by ludralph** [reply]
Hi @ElBaraaRadwan 

Can you add a short delay between starting the database container and deploying the Prisma schema? For example:
```
"db:test:restart": "yarn db:test:rm && yarn db:test:up && sleep 5 && yarn prisma:test:deploy"
```

**Reply by ElBaraaSamir** [reply]
Hi, thank you for replying @ludralph 

I've already tried, and whether it's a short or long delay, it always produces the same error.

**Reply by VuongNhatHien** [reply]
> Hi @ElBaraaRadwan
> 
> Can you add a short delay between starting the database container and deploying the Prisma schema? For example:
> 
> ```
> "db:test:restart": "yarn db:test:rm && yarn db:test:up && sleep 5 && yarn prisma:test:deploy"
> ```

This worked for me, thanks
